### PR TITLE
Add normalize_first arg to Transformer Layers

### LINKED ISSUE #337: Investigate recreating a GPT-2 forward pass with KerasNLP
With the addition of a roughly compatible tokenizer with gpt-2 in https://github.com/keras-team/keras-nlp/pull/303, it should be possible to start experimenting with recreating an entire gpt forward pass with KerasNLP components.

This should be more an investigation and validation for now, as we continue to answer some core design questions with our BERT modeling code.

Rough flow:

 - Load a reference version of the gpt-2 tokenizer and network, either from the [released github](https://github.com/openai/gpt-2) or huggingface.
 - Recreate the gpt-2 tokenizer using https://github.com/keras-team/keras-nlp/pull/303
 - Recreate the decoder models itself using KerasNLP components, and convert weights.
 - Verify that our output can match with a high degree of accuracy.

### PR DESCRIPTION
**Pull request by abheesht17**
Partially resolves #337 

For the `TransformerDecoder` layer, I went ahead and normalized the input to the cross-attention layer as well (other than self-attention and feedforward layers). Considering `normalize_first` for the self-attention layer and feedforward layer and not considering it for the cross-attention layer seemed incomplete.

Some references for the above:

- https://fossies.org/linux/tensorflow-official-models/official/projects/detr/modeling/transformer.py (LN 818-819)
- PyTorch (check `class TransformerDecoderLayer`): https://pytorch.org/docs/stable/_modules/torch/nn/modules/transformer.html#TransformerDecoderLayer
```
        if self.norm_first:
            x = x + self._sa_block(self.norm1(x), tgt_mask, tgt_key_padding_mask)
            x = x + self._mha_block(self.norm2(x), memory, memory_mask, memory_key_padding_mask)
```
- https://github.com/seetaresearch/dragon/blob/494774d3a545f807d483fd9e6e4563cedec6dda5/torch/core/nn/modules/transformer.py#L183-L184

### CODE REVIEW

File: keras_nlp/layers/transformer_decoder.py [hunk 1]
```diff
@@ -266,6 +272,11 @@ def call(
         else:
             decoder_mask = tf.minimum(decoder_mask, causal_mask)
 
+        if self.normalize_first:
```
Comment (mattdangerw) [inline keras_nlp/layers/transformer_decoder.py:275]: We should not have variables that only exist inside concurrent if blocks, that's confusing. We should probably have something more like.

```
input_decoder_sequence = decoder_sequence
if self.normalize_first:
    decoder_sequence = self._decoder_attention_layernorm(decoder_sequence)

...

attention_output = input_decoder_sequence + self_attended
if not self.normalize_first:
    attention_output = self._decoder_attention_layernorm(attention_output)
```

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Looks good! A few comments.

**Comment by mattdangerw** [comment]
Oh also, sounds good re cross attn layer norm and norm first. Agree that is the correct approach.

File: keras_nlp/layers/transformer_decoder.py [hunk 2]
```diff
@@ -63,6 +63,10 @@ class TransformerDecoder(keras.layers.Layer):
         bias_initializer: string or `keras.initializers` initializer,
             defaults to "zeros". The bias initializer for
             the dense and multiheaded attention layers.
+        normalize_first: bool. If True, the inputs to the attention layer(s) and
+            the intermediate dense layer are normalized (similar to GPT-2). If
+            set to False, outputs of attention layer and intermediate dense
+            layer are normalized (similar to BERT). Defaults to False.
```
Comment (mattdangerw) [inline keras_nlp/layers/transformer_decoder.py:69]: Move the "Defaults to False." right after "bool.", that will match the other other args in this file.

File: keras_nlp/layers/transformer_decoder.py [hunk 3]
```diff
@@ -292,19 +307,23 @@ def call(
...
-
+            attention_output = residual_attention_output + cross_attended
+            if not self.normalize_first:
+                attention_output = self._cross_attention_layernorm(
+                    attention_output
+                )
+
+        residual_attention_output = attention_output
+        if self.normalize_first:
+            attention_output = self._feedforward_layernorm(attention_output)
```
Comment (mattdangerw) [inline keras_nlp/layers/transformer_decoder.py:318]: feedforward -> feed_forward in this name to match others

File: keras_nlp/layers/transformer_encoder.py [hunk 4]
```diff
@@ -51,6 +51,10 @@ class TransformerEncoder(keras.layers.Layer):
         bias_initializer: string or `keras.initializers` initializer,
             defaults to "zeros". The bias initializer for
             the dense and multiheaded attention layers.
+        normalize_first: bool. If True, the inputs to the attention layer and
```
Comment (mattdangerw) [inline keras_nlp/layers/transformer_encoder.py:54]: same here, default sentence after bool sentence

**Review by mattdangerw — APPROVED** [review]
lgtm! just a few last nits

File: keras_nlp/layers/transformer_decoder.py (hunk 3 above)
Comment (abheesht17) [reply keras_nlp/layers/transformer_decoder.py:318]: "Feedforward" is one word. So, I'll change `feed_forward` to `feedforward`.

Comment (mattdangerw) [reply keras_nlp/layers/transformer_decoder.py:318]: Sg! As long as we are consistent (preferably across decoder and encoder).

**Comment by mattdangerw** [comment]
Thanks!
